### PR TITLE
feat: add MFA (Phone) related config

### PR DESCRIPTION
**Pull request by J0**
## What kind of change does this PR introduce?

Introduces the supporting Auth configuration for  MFA (Phone). This include:

1. The respective enable / disable flows for each toggle `MFA.Phone.EnrollEnabled` and `MFA.Phone.VerifyEnabled`
2. `MFA.MaxEnrolledFactors`  which is already exposed on the platform
4. Phone also comes with a configurable OTP Length as well as Template. We don't include the provider specific configuration as they are derived from the primary provider. We also expose MaxFrequency

### CODE REVIEW

File: internal/start/start.go [hunk 1]
```diff
@@ -609,6 +609,20 @@ EOF
 				"GOTRUE_HOOK_SEND_EMAIL_SECRETS="+utils.Config.Auth.Hook.SendEmail.Secrets,
 			)
 		}
+		if utils.Config.Auth.MFA.Phone.EnrollEnabled && utils.Config.Auth.MFA.Phone.VerifyEnabled {
+			env = append(
+				env,
+				"GOTRUE_MFA_PHONE_ENROLL_ENABLED=true",
```
Comment (J0) [inline internal/start/start.go:615]:
```suggestion
				"GOTRUE_MFA_PHONE_ENROLL_ENABLED=false",
```

Comment (hf) [reply internal/start/start.go:615]: Yup this should be false

**Review by hf — COMMENTED** [review]
Seems fine.

**Comment by J0** [comment]
How this was tested:

Place this set of entries in the `config.toml`

```
[auth.mfa.totp]
enroll_enabled = true
verify_enabled = true

[auth.mfa]
max_enrolled_factors = 20

[auth.mfa.phone]
enroll_enabled = true
verify_enabled = true
otp_length = 6
template = "Your Code is {{ .Code }}"
max_frequency = "10s"
```

Run `docker inspect` on the Auth container and check that the Auth config is appropriately set

File: pkg/config/config.go [hunk 2]
```diff
@@ -293,6 +294,27 @@ type (
...
+		VerifyEnabled bool `toml:"verify_enabled"`
+	}
+	totpFactorTypeConfiguration struct {
+		factorTypeConfiguration
+	}
+
+	phoneFactorTypeConfiguration struct {
+		factorTypeConfiguration
+		OtpLength    int           `toml:"otp_length"`
+		Template     string        `toml:"template"`
```
Comment (sweatybridge) [inline pkg/config/config.go:308]: Is this referring to a template string or path to a template file?

File: pkg/config/config.go [hunk 3]
```diff
@@ -293,6 +294,27 @@ type (
...
+		EnrollEnabled bool `toml:"enroll_enabled"`
+		VerifyEnabled bool `toml:"verify_enabled"`
+	}
+	totpFactorTypeConfiguration struct {
+		factorTypeConfiguration
+	}
+
+	phoneFactorTypeConfiguration struct {
+		factorTypeConfiguration
+		OtpLength    int           `toml:"otp_length"`
```
Comment (sweatybridge) [inline pkg/config/config.go:307]:
```suggestion
		OtpLength    uint          `toml:"otp_length"`
```

File: pkg/config/config.go [hunk 4]
```diff
@@ -293,6 +294,27 @@ type (
...
+		Template     string        `toml:"template"`
+		MaxFrequency time.Duration `toml:"max_frequency"`
+	}
+
+	mfa struct {
+		TOTP  totpFactorTypeConfiguration  `toml:"totp"`
+		Phone phoneFactorTypeConfiguration `toml:"phone"`
+		// set to float64 for backward compatibility
+		MaxEnrolledFactors float64 `toml:"max_enrolled_factors"`
```
Comment (sweatybridge) [inline pkg/config/config.go:316]: Do we need to consider backwards compatibility if it's a new config? probably use `uint` or the correct type.

Comment (J0) [reply pkg/config/config.go:316]: Yeah good point - the CLI config doesn't need to be compatible against the underlying env config. Let's change

File: pkg/config/config.go (hunk 2 above)
Comment (J0) [reply pkg/config/config.go:308]: This should refer to a Template string. Perhaps can add a comment?

Can change the name of course but the config name in Auth is `Template` so hoping not to deviate.

Open to other suggestions as well

File: pkg/config/config.go [hunk 5]
```diff
@@ -294,6 +295,27 @@ type (
...
+		OtpLength    uint          `toml:"otp_length"`
+		Template     string        `toml:"template"`
+		MaxFrequency time.Duration `toml:"max_frequency"`
+	}
+
+	mfa struct {
+		TOTP  totpFactorTypeConfiguration  `toml:"totp"`
+		Phone phoneFactorTypeConfiguration `toml:"phone"`
+		// set to float64 for backward compatibility
+		MaxEnrolledFactors float64 `toml:"max_enrolled_factors"`
```
Comment (J0) [inline pkg/config/config.go:317]:
```suggestion
		MaxEnrolledFactors uint `toml:"max_enrolled_factors"`
```

**Comment by J0** [comment]
Deploy has completed

File: pkg/config/config.go [hunk 6]
```diff
@@ -294,6 +295,27 @@ type (
 		SendSMS                     hookConfig `toml:"send_sms"`
 		SendEmail                   hookConfig `toml:"send_email"`
 	}
+	factorTypeConfiguration struct {
+		EnrollEnabled bool `toml:"enroll_enabled"`
+		VerifyEnabled bool `toml:"verify_enabled"`
+	}
+	totpFactorTypeConfiguration struct {
+		factorTypeConfiguration
+	}
```
Comment (sweatybridge) [inline pkg/config/config.go:304]:
```suggestion
```

Let's defer adding this type until we actually need to extend it?

File: pkg/config/config.go [hunk 7]
```diff
@@ -294,6 +295,27 @@ type (
...
+
+	phoneFactorTypeConfiguration struct {
+		factorTypeConfiguration
+		OtpLength    uint          `toml:"otp_length"`
+		Template     string        `toml:"template"`
+		MaxFrequency time.Duration `toml:"max_frequency"`
+	}
+
+	mfa struct {
+		TOTP  totpFactorTypeConfiguration  `toml:"totp"`
```
Comment (sweatybridge) [inline pkg/config/config.go:314]:
```suggestion
		TOTP  factorTypeConfiguration       `toml:"totp"`
```

File: pkg/config/config.go (hunk 2 above)
Comment (sweatybridge) [reply pkg/config/config.go:309]: Always good to add an example entry to [config.toml](https://github.com/supabase/cli/blob/develop/pkg/config/templates/config.toml#L116), for eg.

```
[auth.mfa.phone]
enroll_enabled = true
...
```

**Review by sweatybridge — APPROVED** [review]

**Pull request merged**